interface/reports/patient_latest_encounters.php

```diff
@@ -0,0 +1,387 @@
+    $sqlArrayBind = array();
+    $query = "SELECT " .
+    "p.fname, p.mname, p.lname, " .
+    "p.pid, p.pubpid, p.DOB, p.sex, " .
+    "count(e.date) AS ecount, max(e.date) AS edate, " .
+    "i1.date AS idate1, i2.date AS idate2, " .
+    "c1.name AS cname1, c2.name AS cname2 " .
+    "FROM patient_data AS p ";
+
+    if (!empty($from_date)) {
+        $query .= "JOIN form_encounter AS e ON " .
+        "e.pid = p.pid AND " .
+        "e.date >= ? AND " .
+        "e.date <= ? ";
+        array_push($sqlArrayBind, $from_date . ' 00:00:00', $to_date . ' 23:59:59');
+        if ($form_provider) {
+            $query .= "AND e.provider_id = ? ";
+            array_push($sqlArrayBind, $form_provider);
+        }
+    } else {
+        if ($form_provider) {
+            $query .= "JOIN form_encounter AS e ON " .
+            "e.pid = p.pid AND e.provider_id = ? ";
+            array_push($sqlArrayBind, $form_provider);
+        } else {
+            $query .= "LEFT OUTER JOIN form_encounter AS e ON " .
+            "e.pid = p.pid ";
+        }
+    }
+
+    $query .=
+    "LEFT OUTER JOIN insurance_data AS i1 ON " .
+    "i1.pid = p.pid AND i1.type = 'primary' " .
+    "LEFT OUTER JOIN insurance_companies AS c1 ON " .
+    "c1.id = i1.provider " .
+    "LEFT OUTER JOIN insurance_data AS i2 ON " .
+    "i2.pid = p.pid AND i2.type = 'secondary' " .
+    "LEFT OUTER JOIN insurance_companies AS c2 ON " .
+    "c2.id = i2.provider " .
+    "GROUP BY p.lname, p.fname, p.mname, p.pid, i1.date, i2.date " .
+    "ORDER BY p.lname, p.fname, p.mname, p.pid, i1.date DESC, i2.date DESC";
+    $res = sqlStatement($query, $sqlArrayBind);
+
+    $prevpid = 0;
+    while ($row = sqlFetchArray($res)) {
+        if ($row['pid'] == $prevpid) {
+            continue;
+        }
+        $prevpid = $row['pid'];
+        $age = '';
+        if (!empty($row['DOB'])) {
+            $dob = $row['DOB'];
+            $tdy = $row['edate'] ? $row['edate'] : date('Y-m-d');
+            $ageInMonths = (substr($tdy, 0, 4) * 12) + substr($tdy, 5, 2) -
+                   (substr($dob, 0, 4) * 12) - substr($dob, 5, 2);
+            $dayDiff = substr($tdy, 8, 2) - substr($dob, 8, 2);
+            if ($dayDiff < 0) {
+                --$ageInMonths;
+            }
+
+            $age = intval($ageInMonths / 12);
+        }
+        $sqlArrayBind = array();
+        $en_pid = $row['pid'];
+        $equery = "SELECT " .
+            "facility, provider_id, reason, date " .
+            "FROM form_encounter " .
+            "WHERE pid = $en_pid ";
+        $eres = sqlStatement($equery, $sqlArrayBind);
+        // get encounter on the most recent date
+        while ($erow = sqlFetchArray($eres)) {
+            if ($erow['date'] == $row['edate']) {
+                break;
+            }
+        }
+        // get provider name
+        $sqlArrayBind = array();
+        $providerID = $erow['provider_id'];
+        $pquery = "SELECT " . "fname, lname FROM users WHERE id = $providerID";
```

Comment:
need to use sql binding to prevent sql injection attacks

Reply:
Hi @bradymiller - I'm not sure how to do an sql binding, would you be able to give me the code for this search ? (then i can apply the same technique to the previous search) - i've searched around for help but am a little confused. I thought  putting the query into a string ($pquery) and then passing the string to sqlStatement($pquery, $sqlArrayBind)  was doing the binding. many thanks

Reply:
$sqlArrayBind = array();
$providerID = $erow['provider_id'];
$sqlArrayBind[] = $providerID;
$pquery = "SELECT " . "fname, lname FROM users WHERE id = ?";
$pres = sqlStatement($pquery, $sqlArrayBind);
$prow = sqlFetchArray($pres);

The ADODB API will substitute the bindings and pass them in. Order is important from left to right if more than one binding. Essentially add a ? wherever a value is set in the statement and add to the binding array in same order as read by sql engine.

Could also be  $pres = sqlStatement($pquery, array($providerID));

Reply:
many thanks. have made the changes and done new PR